crates/biome_css_semantic/src/semantic_model/builder.rs

```diff
@@ -53,20 +64,40 @@ impl SemanticModelBuilder {
     pub fn push_event(&mut self, event: SemanticEvent) {
         match event {
             SemanticEvent::RuleStart(range) => {
+                let new_rule_id = self.next_rule_id;
+                self.next_rule_id = RuleId::new(new_rule_id.index() + 1);
+
+                let parent_id = self.current_rule_stack.last().copied();
+
                 let new_rule = Rule {
+                    id: new_rule_id,
                     selectors: Vec::new(),
                     declarations: Vec::new(),
-                    children: Vec::new(),
                     range,
+                    parent_id,
+                    child_ids: Vec::new(),
                 };
-                self.current_rule_stack.push(new_rule);
+
+                if let Some(&parent_id) = self.current_rule_stack.last() {
+                    if let Some(parent_rule) = self.rules_by_id.get_mut(&parent_id) {
+                        parent_rule.child_ids.push(new_rule_id);
+                    }
+                }
+
+                self.rules_by_id.insert(new_rule_id, new_rule.clone());
+                self.current_rule_stack.push(new_rule_id);
             }
             SemanticEvent::RuleEnd => {
                 if let Some(completed_rule) = self.current_rule_stack.pop() {
-                    if let Some(parent_rule) = self.current_rule_stack.last_mut() {
-                        parent_rule.children.push(completed_rule);
+                    let completed_rule = &self.rules_by_id[&completed_rule];
+
+                    if self.current_rule_stack.last_mut().is_some() {
+                        self.range_to_rule
+                            .insert(completed_rule.range, completed_rule.clone());
                     } else {
-                        self.rules.push(completed_rule);
+                        self.range_to_rule
+                            .insert(completed_rule.range, completed_rule.clone());
+                        self.rules.push(completed_rule.clone());
```

Comment:
Adding comment of:

- range_to_rule
- descrption of `range_to_rule` handling

will update the readability IMHO.